mlpack-3.1.1 fails to compile with Ensmallen 2.1.0 release

<!--

Welcome!  Please fill out the template below; that makes it easier for us to 
quickly figure out what the issue is and solve it.  Thanks!

-->

#### Issue description

When trying to use build mlpack 3.1.1 with Ensmallen 2.1.0, I receive the following compiler error:

```
/mlpack-3.1.1/src/mlpack/methods/matrix_completion/matrix_completion.cpp: In member function 'void mlpack::matrix_completion::MatrixCompl
etion::Recover(arma::mat&)':
/mlpack-3.1.1/src/mlpack/methods/matrix_completion/matrix_completion.cpp:95:46: error: no matching function for call to 'ens::LRSDPFuncti
on<ens::SDP<arma::SpMat<double> > >::GetInitialPoint()'
   recovered = sdp.Function().GetInitialPoint();
                                              ^
In file included from /usr/local/include/ensmallen_bits/sdp/lrsdp.hpp:18:0,
                 from /usr/local/include/ensmallen.hpp:104,
                 from /mlpack-3.1.1/src/mlpack/methods/matrix_completion/matrix_completion.hpp:16,
                 from /mlpack-3.1.1/src/mlpack/methods/matrix_completion/matrix_completion.cpp:13:
/usr/local/include/ensmallen_bits/sdp/lrsdp_function.hpp:97:18: note: candidate: template<class MatType> const MatType& ens::LRSDPFunctio
n<SDPType>::GetInitialPoint() const [with MatType = MatType; SDPType = ens::SDP<arma::SpMat<double> >]
   const MatType& GetInitialPoint() const
                  ^
/usr/local/include/ensmallen_bits/sdp/lrsdp_function.hpp:97:18: note:   template argument deduction/substitution failed:
/mlpack-3.1.1/src/mlpack/methods/matrix_completion/matrix_completion.cpp:95:46: note:   couldn't deduce template parameter 'MatType'
   recovered = sdp.Function().GetInitialPoint();
                                              ^
[ 29%] Building CXX object src/mlpack/CMakeFiles/mlpack.dir/methods/neighbor_search/unmap.cpp.o
src/mlpack/CMakeFiles/mlpack.dir/build.make:927: recipe for target 'src/mlpack/CMakeFiles/mlpack.dir/methods/matrix_completion/matrix_com
pletion.cpp.o' failed
make[2]: *** [src/mlpack/CMakeFiles/mlpack.dir/methods/matrix_completion/matrix_completion.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:301: recipe for target 'src/mlpack/CMakeFiles/mlpack.dir/all' failed
make[1]: *** [src/mlpack/CMakeFiles/mlpack.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
The command '/bin/bash -c cd mlpack* && mkdir build && cd build     && cmake -DDEBUG=OFF -DPROFILE=OFF -DBUILD_PYTHON_BINDINGS=OFF -DBUIL
D_TESTS=OFF ..     && make -j2 && make install' returned a non-zero code: 2
```

#### Your environment

 * version of mlpack: 3.1.1
 * operating system: Ubuntu 16.04
 * compiler: gcc 5.4.0
 * version of dependencies (Boost/Armadillo): (1.58.0, 9.700.2)
 * any other environment information you think is relevant:
`Dockerfile`:

```
FROM ubuntu:xenial

# USE BASH
SHELL ["/bin/bash", "-c"]

# INSTALL SYSTEM-WIDE DEPENDENCIES
RUN apt-get update && apt-get install -y --fix-missing cmake pkg-config git wget \
    build-essential libboost-all-dev zip
RUN apt-get install -y --allow-unauthenticated libomp-dev libopenblas-dev liblapack-dev \
    libarpack++2-dev

# INSTALL ARMADILLO
RUN wget http://sourceforge.net/projects/arma/files/armadillo-9.700.2.tar.xz \
    && tar -xf armadillo-9.700.2.tar.xz
RUN cd armadillo* && mkdir build && cd build \
    && cmake .. && make -j2 && make install

# INSTALL ENSMALLEN
#RUN wget http://ensmallen.org/files/ensmallen-1.15.1.tar.gz \
#    && tar -xzf ensmallen-1.15.1.tar.gz
RUN wget https://github.com/mlpack/ensmallen/archive/2.10.0.zip \
    && unzip 2.10.0
RUN cd ensmallen* && mkdir build && cd build \
    && cmake -DCMAKE_CXX_FLAGS="-Werror" -DCMAKE_C_FLAGS="-Werror" .. \
    && make -j2 && make install

# INSTALL MLPACK
RUN wget https://www.mlpack.org/files/mlpack-3.1.1.tar.gz \
    && tar -xzf mlpack-3.1.1.tar.gz
RUN cd mlpack* && mkdir build && cd build \
    && cmake -DDEBUG=OFF -DPROFILE=OFF -DBUILD_PYTHON_BINDINGS=OFF -DBUILD_TESTS=OFF .. \
    && make -j2 && make install

# INSTALL NLOHMANN_JSON (downloaded from github)
RUN wget https://github.com/nlohmann/json/archive/release/3.7.0.zip \
    && unzip 3.7.0.zip
RUN cd json* && mkdir build && cd build \
    && cmake .. \
    && make -j2 && make install

# INSTALL OR-TOOLS (from binary)
RUN apt-get install -y zlib1g-dev
RUN wget https://github.com/google/or-tools/releases/download/v7.3/or-tools_ubuntu-16.04_v7.3.7083.tar.gz \
    && tar -xvf or-tools_ubuntu-16.04_v7.3.7083.tar.gz
ENV ORTOOLS_ROOT=/or-tools_Ubuntu-16.04-64bit_v7.3.7083

# REMOVE ARCHIVES
RUN rm *.tar.* && rm *.zip
```

#### Steps to reproduce

`docker build --network=host -t image_name .` from directory with `Dockerfile` above in it.

#### Expected behavior

The image build should complete successfully.

#### Actual behavior

Image failed to build; see original error for info.

--- a/.travis.yml
+++ b/.travis.yml
@@ -41,7 +41,7 @@ matrix:
         - brew install openblas armadillo || brew install openblas armadillo
 
 install:
-  - mkdir build && cd build && cmake $CMAKE_OPTIONS .. && travis_wait 60 make -j2
+  - mkdir build && cd build && cmake $CMAKE_OPTIONS .. && travis_wait 75 make -j2
 
 script:
   - CTEST_OUTPUT_ON_FAILURE=1 travis_wait 30 ctest -j2

--- a/src/mlpack/methods/reinforcement_learning/q_learning.hpp
+++ b/src/mlpack/methods/reinforcement_learning/q_learning.hpp
@@ -85,6 +85,11 @@ class QLearning
             EnvironmentType environment = EnvironmentType());
 
   /**
+   * Clean memory.
+   */
+  ~QLearning();
+
+  /**
    * Execute a step in an episode.
    * @return Reward for the step.
    */
@@ -140,6 +145,9 @@ class QLearning
 
   //! Locally-stored updater.
   UpdaterType updater;
+  #if ENS_VERSION_MAJOR >= 2
+  typename UpdaterType::template Policy<arma::mat, arma::mat>* updatePolicy;
+  #endif
 
   //! Locally-stored behavior policy.
   PolicyType policy;

--- a/src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp
+++ b/src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp
@@ -39,6 +39,9 @@ QLearning<
     config(std::move(config)),
     learningNetwork(std::move(network)),
     updater(std::move(updater)),
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy(NULL),
+    #endif
     policy(std::move(policy)),
     replayMethod(std::move(replayMethod)),
     environment(std::move(environment)),
@@ -49,9 +52,37 @@ QLearning<
   if (learningNetwork.Parameters().is_empty())
     learningNetwork.ResetParameters();
 
+  #if ENS_VERSION_MAJOR == 1
   this->updater.Initialize(learningNetwork.Parameters().n_rows,
-      learningNetwork.Parameters().n_cols);
+                           learningNetwork.Parameters().n_cols);
+  #else
+  this->updatePolicy = new typename UpdaterType::template
+      Policy<arma::mat, arma::mat>(this->updater,
+                                   learningNetwork.Parameters().n_rows,
+                                   learningNetwork.Parameters().n_cols);
+  #endif
+
   targetNetwork = learningNetwork;
+}
+
+template <
+  typename EnvironmentType,
+  typename NetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+QLearning<
+  EnvironmentType,
+  NetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::~QLearning()
+{
+  #if ENS_VERSION_MAJOR >= 2
+  delete updatePolicy;
+  #endif
 }
 
 template <
@@ -166,14 +197,18 @@ double QLearning<
     }
   }
 
-  // Learn form experience.
+  // Learn from experience.
   arma::mat gradients;
   learningNetwork.Backward(target, gradients);
 
-  replayMethod.Update(target, sampledActions,
-      nextActionValues, gradients);
+  replayMethod.Update(target, sampledActions, nextActionValues, gradients);
 
+  #if ENS_VERSION_MAJOR == 1
   updater.Update(learningNetwork.Parameters(), config.StepSize(), gradients);
+  #else
+  updatePolicy->Update(learningNetwork.Parameters(), config.StepSize(),
+      gradients);
+  #endif
 
   return reward;
 }

--- a/src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp
+++ b/src/mlpack/methods/reinforcement_learning/worker/n_step_q_learning_worker.hpp
@@ -54,6 +54,9 @@ class NStepQLearningWorker
       const TrainingConfig& config,
       bool deterministic):
       updater(updater),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
       environment(environment),
       config(config),
       deterministic(deterministic),
@@ -61,13 +64,167 @@ class NStepQLearningWorker
   { Reset(); }
 
   /**
+   * Copy another NStepQLearningWorker.
+   *
+   * @param other NStepQLearningWorker to copy.
+   */
+  NStepQLearningWorker(const NStepQLearningWorker& other) :
+      updater(other.updater),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
+      environment(other.environment),
+      config(other.config),
+      deterministic(other.deterministic),
+      steps(other.steps),
+      episodeReturn(other.episodeReturn),
+      pending(other.pending),
+      pendingIndex(other.pendingIndex),
+      network(other.network),
+      state(other.state)
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+
+    Reset();
+  }
+
+  /**
+   * Take ownership of another NStepQLearningWorker.
+   *
+   * @param other NStepQLearningWorker to take ownership of.
+   */
+  NStepQLearningWorker(NStepQLearningWorker&& other) :
+      updater(std::move(other.updater)),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
+      environment(std::move(other.environment)),
+      config(std::move(other.config)),
+      deterministic(std::move(other.deterministic)),
+      steps(std::move(other.steps)),
+      episodeReturn(std::move(other.episodeReturn)),
+      pending(std::move(other.pending)),
+      pendingIndex(std::move(other.pendingIndex)),
+      network(std::move(other.network)),
+      state(std::move(other.state))
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    other.updatePolicy = NULL;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+  }
+
+  /**
+   * Copy another NStepQLearningWorker.
+   *
+   * @param other NStepQLearningWorker to copy.
+   */
+  NStepQLearningWorker& operator=(const NStepQLearningWorker& other)
+  {
+    if (&other == this)
+      return *this;
+
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+
+    updater = other.updater;
+    environment = other.environment;
+    config = other.config;
+    deterministic = other.deterministic;
+    steps = other.steps;
+    episodeReturn = other.episodeReturn;
+    pending = other.pending;
+    pendingIndex = other.pendingIndex;
+    network = other.network;
+    state = other.state;
+
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+
+    Reset();
+
+    return *this;
+  }
+
+  /**
+   * Take ownership of another NStepQLearningWorker.
+   *
+   * @param other NStepQLearningWorker to take ownership of.
+   */
+  NStepQLearningWorker& operator=(NStepQLearningWorker&& other)
+  {
+    if (&other == this)
+      return *this;
+
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+
+    updater = std::move(other.updater);
+    environment = std::move(other.environment);
+    config = std::move(other.config);
+    deterministic = std::move(other.deterministic);
+    steps = std::move(other.steps);
+    episodeReturn = std::move(other.episodeReturn);
+    pending = std::move(other.pending);
+    pendingIndex = std::move(other.pendingIndex);
+    network = std::move(other.network);
+    state = std::move(other.state);
+
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+
+    other.updatePolicy = NULL;
+    #endif
+
+    return *this;
+  }
+
+  /**
+   * Clean memory.
+   */
+  ~NStepQLearningWorker()
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+  }
+
+  /**
    * Initialize the worker.
    * @param learningNetwork The shared network.
    */
   void Initialize(NetworkType& learningNetwork)
   {
+    #if ENS_VERSION_MAJOR == 1
     updater.Initialize(learningNetwork.Parameters().n_rows,
-        learningNetwork.Parameters().n_cols);
+                       learningNetwork.Parameters().n_cols);
+    #else
+    delete updatePolicy;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     learningNetwork.Parameters().n_rows,
+                                     learningNetwork.Parameters().n_cols);
+    #endif
+
     // Build local network.
     network = learningNetwork;
   }
@@ -162,8 +319,13 @@ class NStepQLearningWorker
           config.GradientLimit()); });
 
       // Perform async update of the global network.
-      updater.Update(learningNetwork.Parameters(),
+      #if ENS_VERSION_MAJOR == 1
+      updater.Update(learningNetwork.Parameters(), config.StepSize(),
+          totalGradients);
+      #else
+      updatePolicy->Update(learningNetwork.Parameters(),
           config.StepSize(), totalGradients);
+      #endif
 
       // Sync the local network with the global network.
       network = learningNetwork;
@@ -204,6 +366,9 @@ class NStepQLearningWorker
 
   //! Locally-stored optimizer.
   UpdaterType updater;
+  #if ENS_VERSION_MAJOR >= 2
+  typename UpdaterType::template Policy<arma::mat, arma::mat>* updatePolicy;
+  #endif
 
   //! Locally-stored task.
   EnvironmentType environment;

--- a/src/mlpack/methods/reinforcement_learning/worker/one_step_q_learning_worker.hpp
+++ b/src/mlpack/methods/reinforcement_learning/worker/one_step_q_learning_worker.hpp
@@ -54,6 +54,9 @@ class OneStepQLearningWorker
       const TrainingConfig& config,
       bool deterministic):
       updater(updater),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
       environment(environment),
       config(config),
       deterministic(deterministic),
@@ -61,13 +64,167 @@ class OneStepQLearningWorker
   { Reset(); }
 
   /**
+   * Copy another OneStepQLearningWorker.
+   *
+   * @param other OneStepQLearningWorker to copy.
+   */
+  OneStepQLearningWorker(const OneStepQLearningWorker& other) :
+      updater(other.updater),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
+      environment(other.environment),
+      config(other.config),
+      deterministic(other.deterministic),
+      steps(other.steps),
+      episodeReturn(other.episodeReturn),
+      pending(other.pending),
+      pendingIndex(other.pendingIndex),
+      network(other.network),
+      state(other.state)
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+
+    Reset();
+  }
+
+  /**
+   * Take ownership of another OneStepQLearningWorker.
+   *
+   * @param other OneStepQLearningWorker to take ownership of.
+   */
+  OneStepQLearningWorker(OneStepQLearningWorker&& other) :
+      updater(std::move(other.updater)),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
+      environment(std::move(other.environment)),
+      config(std::move(other.config)),
+      deterministic(std::move(other.deterministic)),
+      steps(std::move(other.steps)),
+      episodeReturn(std::move(other.episodeReturn)),
+      pending(std::move(other.pending)),
+      pendingIndex(std::move(other.pendingIndex)),
+      network(std::move(other.network)),
+      state(std::move(other.state))
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    other.updatePolicy = NULL;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+  }
+
+  /**
+   * Copy another OneStepQLearningWorker.
+   *
+   * @param other OneStepQLearningWorker to copy.
+   */
+  OneStepQLearningWorker& operator=(const OneStepQLearningWorker& other)
+  {
+    if (&other == this)
+      return *this;
+
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+
+    updater = other.updater;
+    environment = other.environment;
+    config = other.config;
+    deterministic = other.deterministic;
+    steps = other.steps;
+    episodeReturn = other.episodeReturn;
+    pending = other.pending;
+    pendingIndex = other.pendingIndex;
+    network = other.network;
+    state = other.state;
+
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+
+    Reset();
+
+    return *this;
+  }
+
+  /**
+   * Take ownership of another OneStepQLearningWorker.
+   *
+   * @param other OneStepQLearningWorker to take ownership of.
+   */
+  OneStepQLearningWorker& operator=(OneStepQLearningWorker&& other)
+  {
+    if (&other == this)
+      return *this;
+
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+
+    updater = std::move(other.updater);
+    environment = std::move(other.environment);
+    config = std::move(other.config);
+    deterministic = std::move(other.deterministic);
+    steps = std::move(other.steps);
+    episodeReturn = std::move(other.episodeReturn);
+    pending = std::move(other.pending);
+    pendingIndex = std::move(other.pendingIndex);
+    network = std::move(other.network);
+    state = std::move(other.state);
+
+    #if ENS_VERSION_MAJOR >= 2
+    other.updatePolicy = NULL;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+
+    return *this;
+  }
+
+  /**
+   * Clean memory.
+   */
+  ~OneStepQLearningWorker()
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+  }
+
+  /**
    * Initialize the worker.
    * @param learningNetwork The shared network.
    */
   void Initialize(NetworkType& learningNetwork)
   {
+    #if ENS_VERSION_MAJOR == 1
     updater.Initialize(learningNetwork.Parameters().n_rows,
-        learningNetwork.Parameters().n_cols);
+                       learningNetwork.Parameters().n_cols);
+    #else
+    delete updatePolicy;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     learningNetwork.Parameters().n_rows,
+                                     learningNetwork.Parameters().n_cols);
+    #endif
+
     // Build local network.
     network = learningNetwork;
   }
@@ -162,8 +319,13 @@ class OneStepQLearningWorker
           config.GradientLimit()); });
 
       // Perform async update of the global network.
-      updater.Update(learningNetwork.Parameters(),
+      #if ENS_VERSION_MAJOR == 1
+      updater.Update(learningNetwork.Parameters(), config.StepSize(),
+          totalGradients);
+      #else
+      updatePolicy->Update(learningNetwork.Parameters(),
           config.StepSize(), totalGradients);
+      #endif
 
       // Sync the local network with the global network.
       network = learningNetwork;
@@ -204,6 +366,9 @@ class OneStepQLearningWorker
 
   //! Locally-stored optimizer.
   UpdaterType updater;
+  #if ENS_VERSION_MAJOR >= 2
+  typename UpdaterType::template Policy<arma::mat, arma::mat>* updatePolicy;
+  #endif
 
   //! Locally-stored task.
   EnvironmentType environment;

--- a/src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp
+++ b/src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp
@@ -82,7 +82,8 @@ class OneStepSarsaWorker
       pending(other.pending),
       pendingIndex(other.pendingIndex),
       network(other.network),
-      state(other.state)
+      state(other.state),
+      action(other.action)
   {
     Reset();
 
@@ -112,7 +113,8 @@ class OneStepSarsaWorker
       pending(std::move(other.pending)),
       pendingIndex(std::move(other.pendingIndex)),
       network(std::move(other.network)),
-      state(std::move(other.state))
+      state(std::move(other.state)),
+      action(std::move(other.action))
   {
     #if ENS_VERSION_MAJOR >= 2
     other.updatePolicy = NULL;
@@ -148,6 +150,7 @@ class OneStepSarsaWorker
     pendingIndex = other.pendingIndex;
     network = other.network;
     state = other.state;
+    action = other.action;
 
     #if ENS_VERSION_MAJOR >= 2
     updatePolicy = new typename UpdaterType::template
@@ -185,6 +188,7 @@ class OneStepSarsaWorker
     pendingIndex = std::move(other.pendingIndex);
     network = std::move(other.network);
     state = std::move(other.state);
+    action = std::move(other.action);
 
     #if ENS_VERSION_MAJOR >= 2
     other.updatePolicy = NULL;

--- a/src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp
+++ b/src/mlpack/methods/reinforcement_learning/worker/one_step_sarsa_worker.hpp
@@ -55,6 +55,9 @@ class OneStepSarsaWorker
       const TrainingConfig& config,
       bool deterministic):
       updater(updater),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
       environment(environment),
       config(config),
       deterministic(deterministic),
@@ -62,13 +65,167 @@ class OneStepSarsaWorker
   { Reset(); }
 
   /**
+   * Copy another OneStepSarsaWorker.
+   *
+   * @param other OneStepSarsaWorker to copy.
+   */
+  OneStepSarsaWorker(const OneStepSarsaWorker& other) :
+      updater(other.updater),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
+      environment(other.environment),
+      config(other.config),
+      deterministic(other.deterministic),
+      steps(other.steps),
+      episodeReturn(other.episodeReturn),
+      pending(other.pending),
+      pendingIndex(other.pendingIndex),
+      network(other.network),
+      state(other.state)
+  {
+    Reset();
+
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+  }
+
+  /**
+   * Take ownership of another OneStepSarsaWorker.
+   *
+   * @param other OneStepSarsaWorker to take ownership of.
+   */
+  OneStepSarsaWorker(OneStepSarsaWorker&& other) :
+      updater(std::move(other.updater)),
+      #if ENS_VERSION_MAJOR >= 2
+      updatePolicy(NULL),
+      #endif
+      environment(std::move(other.environment)),
+      config(std::move(other.config)),
+      deterministic(std::move(other.deterministic)),
+      steps(std::move(other.steps)),
+      episodeReturn(std::move(other.episodeReturn)),
+      pending(std::move(other.pending)),
+      pendingIndex(std::move(other.pendingIndex)),
+      network(std::move(other.network)),
+      state(std::move(other.state))
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    other.updatePolicy = NULL;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+  }
+
+  /**
+   * Copy another OneStepSarsaWorker.
+   *
+   * @param other OneStepSarsaWorker to copy.
+   */
+  OneStepSarsaWorker& operator=(const OneStepSarsaWorker& other)
+  {
+    if (&other == this)
+      return *this;
+
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+
+    updater = other.updater;
+    environment = other.environment;
+    config = other.config;
+    deterministic = other.deterministic;
+    steps = other.steps;
+    episodeReturn = other.episodeReturn;
+    pending = other.pending;
+    pendingIndex = other.pendingIndex;
+    network = other.network;
+    state = other.state;
+
+    #if ENS_VERSION_MAJOR >= 2
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+
+    Reset();
+
+    return *this;
+  }
+
+  /**
+   * Take ownership of another OneStepSarsaWorker.
+   *
+   * @param other OneStepSarsaWorker to take ownership of.
+   */
+  OneStepSarsaWorker& operator=(OneStepSarsaWorker&& other)
+  {
+    if (&other == this)
+      return *this;
+
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+
+    updater = std::move(other.updater);
+    environment = std::move(other.environment);
+    config = std::move(other.config);
+    deterministic = std::move(other.deterministic);
+    steps = std::move(other.steps);
+    episodeReturn = std::move(other.episodeReturn);
+    pending = std::move(other.pending);
+    pendingIndex = std::move(other.pendingIndex);
+    network = std::move(other.network);
+    state = std::move(other.state);
+
+    #if ENS_VERSION_MAJOR >= 2
+    other.updatePolicy = NULL;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     network.Parameters().n_rows,
+                                     network.Parameters().n_cols);
+    #endif
+
+    return *this;
+  }
+
+  /**
+   * Clean memory.
+   */
+  ~OneStepSarsaWorker()
+  {
+    #if ENS_VERSION_MAJOR >= 2
+    delete updatePolicy;
+    #endif
+  }
+
+  /**
    * Initialize the worker.
    * @param learningNetwork The shared network.
    */
   void Initialize(NetworkType& learningNetwork)
   {
+    #if ENS_VERSION_MAJOR == 1
     updater.Initialize(learningNetwork.Parameters().n_rows,
-        learningNetwork.Parameters().n_cols);
+                       learningNetwork.Parameters().n_cols);
+    #else
+    delete updatePolicy;
+
+    updatePolicy = new typename UpdaterType::template
+        Policy<arma::mat, arma::mat>(updater,
+                                     learningNetwork.Parameters().n_rows,
+                                     learningNetwork.Parameters().n_cols);
+    #endif
+
     // Build local network.
     network = learningNetwork;
   }
@@ -171,8 +328,13 @@ class OneStepSarsaWorker
           config.GradientLimit()); });
 
       // Perform async update of the global network.
-      updater.Update(learningNetwork.Parameters(),
+      #if ENS_VERSION_MAJOR == 1
+      updater.Update(learningNetwork.Parameters(), config.StepSize(),
+          totalGradients);
+      #else
+      updatePolicy->Update(learningNetwork.Parameters(),
           config.StepSize(), totalGradients);
+      #endif
 
       // Sync the local network with the global network.
       network = learningNetwork;
@@ -215,6 +377,9 @@ class OneStepSarsaWorker
 
   //! Locally-stored optimizer.
   UpdaterType updater;
+  #if ENS_VERSION_MAJOR >= 2
+  typename UpdaterType::template Policy<arma::mat, arma::mat>* updatePolicy;
+  #endif
 
   //! Locally-stored task.
   EnvironmentType environment;

--- a/src/mlpack/tests/feedforward_network_test.cpp
+++ b/src/mlpack/tests/feedforward_network_test.cpp
@@ -159,7 +159,12 @@ BOOST_AUTO_TEST_CASE(ForwardBackwardTest)
 
   ens::VanillaUpdate opt;
   model.ResetParameters();
+  #if ENS_VERSION_MAJOR == 1
   opt.Initialize(model.Parameters().n_rows, model.Parameters().n_cols);
+  #else
+  ens::VanillaUpdate::Policy<arma::mat, arma::mat> optPolicy(opt,
+      model.Parameters().n_rows, model.Parameters().n_cols);
+  #endif
   double stepSize = 0.01;
   size_t batchSize = 10;
 
@@ -179,7 +184,11 @@ BOOST_AUTO_TEST_CASE(ForwardBackwardTest)
       model.Forward(currentData, currentResuls);
       arma::mat gradients;
       model.Backward(currentLabels, gradients);
+      #if ENS_VERSION_MAJOR == 1
       opt.Update(model.Parameters(), stepSize, gradients);
+      #else
+      optPolicy.Update(model.Parameters(), stepSize, gradients);
+      #endif
       batchStart = batchEnd;
 
       arma::mat prediction = arma::zeros<arma::mat>(1, currentResuls.n_cols);

